sslr-core/src/main/java/com/sonar/sslr/impl/typed/SyntaxTreeCreator.java

```diff
@@ -0,0 +1,193 @@
+
+public class SyntaxTreeCreator<T> {
+
+  private final Object treeFactory;
+  private final GrammarBuilderInterceptor mapping;
+  private final NodeBuilder nodeBuilder;
+
+  private final Token.Builder tokenBuilder = Token.builder();
+  private final List<Trivia> trivias = new ArrayList<>();
+
+  private Input input;
+
+  public SyntaxTreeCreator(Object treeFactory, GrammarBuilderInterceptor mapping, NodeBuilder nodeBuilder) {
+    this.treeFactory = treeFactory;
+    this.mapping = mapping;
+    this.nodeBuilder = nodeBuilder;
+  }
+
+  public T create(ParseNode node, Input input) {
+    this.input = input;
+    this.trivias.clear();
+    T result = (T) visit(node);
+    if (result instanceof AstNode) {
+      ((AstNode) result).hasToBeSkippedFromAst();
+    }
+    return result;
+  }
+
+  private Object visit(ParseNode node) {
+    if (node.getMatcher() instanceof MutableParsingRule) {
+      return visitNonTerminal(node);
+    } else {
+      return visitTerminal(node);
+    }
+  }
+
+  private Object visitNonTerminal(ParseNode node) {
+    MutableParsingRule rule = (MutableParsingRule) node.getMatcher();
+    GrammarRuleKey ruleKey = rule.getRuleKey();
+    Method method = mapping.actionForRuleKey(ruleKey);
+
+    Object result;
+
+    if (mapping.hasMethodForRuleKey(ruleKey)) {
+
+      // TODO Drop useless intermediate nodes
+      Preconditions.checkState(node.getChildren().size() == 1);
+      result = visit(node.getChildren().get(0));
+
+    } else if (mapping.isOptionalRule(ruleKey)) {
+
+      Preconditions.checkState(node.getChildren().size() <= 1);
+      if (node.getChildren().isEmpty()) {
+        result = Optional.absent();
+      } else {
+        Object child = visit(node.getChildren().get(0));
+        if (child instanceof AstNode) {
+          ((AstNode) child).hasToBeSkippedFromAst();
+        }
+        result = Optional.of(child);
+      }
+
+    } else if (mapping.isOneOrMoreRule(ruleKey)) {
+
+      result = convertChildren(node);
+
+    } else if (mapping.isZeroOrMoreRule(ruleKey)) {
+
+      List<Object> convertedChildren = convertChildren(node);
+      result = convertedChildren.isEmpty() ? Optional.absent() : Optional.of(convertedChildren);
+
+    } else if (method == null) {
+
+      result = nodeBuilder.createNonTerminal(rule.getRuleKey(), rule, convertChildren(node), node.getStartIndex(), node.getEndIndex());
+
+    } else {
+
+      List<Object> convertedChildren = convertChildren(node);
+      result = ReflectionUtils.invokeMethod(method, treeFactory, convertedChildren.toArray(new Object[convertedChildren.size()]));
+
+    }
+    return result;
+  }
+
+  private List<Object> convertChildren(ParseNode node) {
+    List<Object> convertedChildren = new ArrayList<>();
+    for (ParseNode child : node.getChildren()) {
+      Object result = visit(child);
+
+      if (result != null) {
+        // FIXME to remove aafter full migration: Allow to skip optional nodes that are supposed to bw skipped from the AST
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)